server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -307,36 +306,11 @@ private synchronized DocumentMapper internalMerge(String type, CompressedXConten
         return internalMerge(documentMapper, reason);
     }
 
-    static void validateTypeName(String type) {
-        if (type.length() == 0) {
-            throw new InvalidTypeNameException("mapping type name is empty");
-        }
-        if (type.length() > 255) {
-            throw new InvalidTypeNameException("mapping type name [" + type + "] is too long; limit is length 255 but was ["
-                + type.length() + "]");
-        }
-        if (type.charAt(0) == '_' && SINGLE_MAPPING_NAME.equals(type) == false) {
-            throw new InvalidTypeNameException("mapping type name [" + type + "] can't start with '_' unless it is called ["
-                + SINGLE_MAPPING_NAME + "]");
-        }
-        if (type.contains("#")) {
-            throw new InvalidTypeNameException("mapping type name [" + type + "] should not include '#' in it");
-        }
-        if (type.contains(",")) {
-            throw new InvalidTypeNameException("mapping type name [" + type + "] should not include ',' in it");
-        }
-        if (type.charAt(0) == '.') {
-            throw new IllegalArgumentException("mapping type name [" + type + "] must not start with a '.'");
-        }
-    }
-
     private synchronized DocumentMapper internalMerge(DocumentMapper mapper, MergeReason reason) {
         boolean hasNested = this.hasNested;
         Map<String, ObjectMapper> fullPathObjectMappers = this.fullPathObjectMappers;
 
         assert mapper != null;
-        // check naming
-        validateTypeName(mapper.type());
```

Comment:
To be safe we could validate here that the type is equal to `_doc`. I think it may still be possible to pass through a custom type name using a transport `CreateIndexRequest`.

Reply:
Unfortunately we can't assume that `type` will be `_doc`, because we could be updating mappings on a 7x index that had a custom type name.  `CreateIndexRequest` asserts on the type name in its `StreamInput` constructor, so I think we should be safe there.

Reply:
Got it, I see the difficulty. I'm okay to merge as-is.